### PR TITLE
Add CI for cuda build

### LINKED ISSUE #67: c++: error: unrecognized command line option '--diag_suppress=esa_on_defaulted_function_ignored'
Cuda fans, any idea about this error:
```
c++: error: unrecognized command line option '--diag_suppress=esa_on_defaulted_function_ignored'
```

I tried a different gcc already....

Details:
```
Running with gitlab-runner  (c364eff5)
  on ascent_copa_setuid 47b77dfd
Using SetUID Shell executor...
Running on login1...
Fetching changes...
Removing kokkos.install/
Removing kokkos/
HEAD is now at 6e6c9eb try newer gcc
Checking out 6e6c9eb9 as ci-cuda...
Skipping Git submodules setup
Downloading artifacts for BuildKokkos Cuda (38102)...
About to register the Shell executor type...
About to register the Batch Runner...
Downloading artifacts from coordinator... ok        id=38102 responseStatus=200 OK token=xheXy1AZ
$ CI_PROJECT_DIR=${PWD}
$ module load cmake
$ module load cuda
$ module load gcc/6.4.0

Lmod is automatically replacing "xl/16.1.1-1" with "gcc/6.4.0".


Due to MODULEPATH changes, the following have been reloaded:
  1) spectrum-mpi/10.2.0.10-20181214

$ for i in ${BACKENDS}; do CMAKE_OPTS+=( -DCabana_ENABLE_${i}=ON ); done
$ j="$(grep -c processor /proc/cpuinfo 2>/dev/null)" || j=0; ((j++))
$ mkdir build && cd build && cmake -DCMAKE_PREFIX_PATH=${CI_PROJECT_DIR}/kokkos.install -DCabana_ENABLE_TESTING=ON -DCabana_ENABLE_Serial=OFF -DCabana_ENABLE_EXAMPLES=OFF ${CMAKE_OPTS[@]} .. && make -k -j${j} VERBOSE=1 && make test CTEST_OUTPUT_ON_FAILURE=1
-- The CXX compiler identification is GNU 6.4.0
-- Check for working CXX compiler: /sw/ascent/gcc/6.4.0/bin/c++
-- Check for working CXX compiler: /sw/ascent/gcc/6.4.0/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found KOKKOS: /root/hpc-gitlab-runner/ecp-copa/cabana/builds/users/junghans/47b77dfd/0/ecp-copa/cabana/kokkos.install  
-- The CUDA compiler identification is NVIDIA 9.2.148
-- Check for working CUDA compiler: /sw/ascent/cuda/9.2.148/bin/nvcc
-- Check for working CUDA compiler: /sw/ascent/cuda/9.2.148/bin/nvcc -- works
-- Detecting CUDA compiler ABI info
-- Detecting CUDA compiler ABI info - done
-- Enable Devices: Cuda
-- Found Doxygen: /bin/doxygen (found version "1.8.5") found components:  doxygen missing components:  dot
-- Performing Test COMPILER_SUPPORTS_MARCH
-- Performing Test COMPILER_SUPPORTS_MARCH - Failed
-- Found Git: /bin/git (found version "1.8.3.1") 
-- Cabana Revision = '6e6c9eb9aa8964c7cb2856f472ce60829eb4aa2d'
-- Configuring done
-- Generating done
-- Build files have been written to: /root/hpc-gitlab-runner/ecp-copa/cabana/builds/users/junghans/47b77dfd/0/ecp-copa/cabana/build
/autofs/nccsopen-svm1_sw/ascent/.swci/0-core/opt/spack/20180914/linux-rhel7-ppc64le/gcc-4.8.5/cmake-3.12.2-vltu2cerbdpbvzwpil2fvw4skvjmabfl/bin/cmake -H/root/hpc-gitlab-runner/ecp-copa/cabana/builds/users/junghans/47b77dfd/0/ecp-copa/cabana -B/root/hpc-gitlab-runner/ecp-copa/cabana/builds/users/junghans/47b77dfd/0/ecp-copa/cabana/build --check-build-system CMakeFiles/Makefile.cmake 0
/autofs/nccsopen-svm1_sw/ascent/.swci/0-core/opt/spack/20180914/linux-rhel7-ppc64le/gcc-4.8.5/cmake-3.12.2-vltu2cerbdpbvzwpil2fvw4skvjmabfl/bin/cmake -E cmake_progress_start /root/hpc-gitlab-runner/ecp-copa/cabana/builds/users/junghans/47b77dfd/0/ecp-copa/cabana/build/CMakeFiles /root/hpc-gitlab-runner/ecp-copa/cabana/builds/users/junghans/47b77dfd/0/ecp-copa/cabana/build/CMakeFiles/progress.marks
make -f CMakeFiles/Makefile2 all
make[1]: Entering directory `/root/hpc-gitlab-runner/ecp-copa/cabana/builds/users/junghans/47b77dfd/0/ecp-copa/cabana/build'
make -f core/src/CMakeFiles/cabanacore.dir/build.make core/src/CMakeFiles/cabanacore.dir/depend
make -f core/unit_test/CMakeFiles/cabana_core_gtest.dir/build.make core/unit_test/CMakeFiles/cabana_core_gtest.dir/depend
make[2]: Entering directory `/root/hpc-gitlab-runner/ecp-copa/cabana/builds/users/junghans/47b77dfd/0/ecp-copa/cabana/build'
cd /root/hpc-gitlab-runner/ecp-copa/cabana/builds/users/junghans/47b77dfd/0/ecp-copa/cabana/build && /autofs/nccsopen-svm1_sw/ascent/.swci/0-core/opt/spack/20180914/linux-rhel7-ppc64le/gcc-4.8.5/cmake-3.12.2-vltu2cerbdpbvzwpil2fvw4skvjmabfl/bin/cmake -E cmake_depends "Unix Makefiles" /root/hpc-gitlab-runner/ecp-copa/cabana/builds/users/junghans/47b77dfd/0/ecp-copa/cabana /root/hpc-gitlab-runner/ecp-copa/cabana/builds/users/junghans/47b77dfd/0/ecp-copa/cabana/core/src /root/hpc-gitlab-runner/ecp-copa/cabana/builds/users/junghans/47b77dfd/0/ecp-copa/cabana/build /root/hpc-gitlab-runner/ecp-copa/cabana/builds/users/junghans/47b77dfd/0/ecp-copa/cabana/build/core/src /root/hpc-gitlab-runner/ecp-copa/cabana/builds/users/junghans/47b77dfd/0/ecp-copa/cabana/build/core/src/CMakeFiles/cabanacore.dir/DependInfo.cmake --color=
make[2]: Entering directory `/root/hpc-gitlab-runner/ecp-copa/cabana/builds/users/junghans/47b77dfd/0/ecp-copa/cabana/build'
cd /root/hpc-gitlab-runner/ecp-copa/cabana/builds/users/junghans/47b77dfd/0/ecp-copa/cabana/build && /autofs/nccsopen-svm1_sw/ascent/.swci/0-core/opt/spack/20180914/linux-rhel7-ppc64le/gcc-4.8.5/cmake-3.12.2-vltu2cerbdpbvzwpil2fvw4skvjmabfl/bin/cmake -E cmake_depends "Unix Makefiles" /root/hpc-gitlab-runner/ecp-copa/cabana/builds/users/junghans/47b77dfd/0/ecp-copa/cabana /root/hpc-gitlab-runner/ecp-copa/cabana/builds/users/junghans/47b77dfd/0/ecp-copa/cabana/core/unit_test /root/hpc-gitlab-runner/ecp-copa/cabana/builds/users/junghans/47b77dfd/0/ecp-copa/cabana/build /root/hpc-gitlab-runner/ecp-copa/cabana/builds/users/junghans/47b77dfd/0/ecp-copa/cabana/build/core/unit_test /root/hpc-gitlab-runner/ecp-copa/cabana/builds/users/junghans/47b77dfd/0/ecp-copa/cabana/build/core/unit_test/CMakeFiles/cabana_core_gtest.dir/DependInfo.cmake --color=
Dependee "/root/hpc-gitlab-runner/ecp-copa/cabana/builds/users/junghans/47b77dfd/0/ecp-copa/cabana/build/core/unit_test/CMakeFiles/cabana_core_gtest.dir/DependInfo.cmake" is newer than depender "/root/hpc-gitlab-runner/ecp-copa/cabana/builds/users/junghans/47b77dfd/0/ecp-copa/cabana/build/core/unit_test/CMakeFiles/cabana_core_gtest.dir/depend.internal".
Dependee "/root/hpc-gitlab-runner/ecp-copa/cabana/builds/users/junghans/47b77dfd/0/ecp-copa/cabana/build/core/src/CMakeFiles/cabanacore.dir/DependInfo.cmake" is newer than depender "/root/hpc-gitlab-runner/ecp-copa/cabana/builds/users/junghans/47b77dfd/0/ecp-copa/cabana/build/core/src/CMakeFiles/cabanacore.dir/depend.internal".
Dependee "/root/hpc-gitlab-runner/ecp-copa/cabana/builds/users/junghans/47b77dfd/0/ecp-copa/cabana/build/core/unit_test/CMakeFiles/CMakeDirectoryInformation.cmake" is newer than depender "/root/hpc-gitlab-runner/ecp-copa/cabana/builds/users/junghans/47b77dfd/0/ecp-copa/cabana/build/core/unit_test/CMakeFiles/cabana_core_gtest.dir/depend.internal".
Dependee "/root/hpc-gitlab-runner/ecp-copa/cabana/builds/users/junghans/47b77dfd/0/ecp-copa/cabana/build/core/src/CMakeFiles/CMakeDirectoryInformation.cmake" is newer than depender "/root/hpc-gitlab-runner/ecp-copa/cabana/builds/users/junghans/47b77dfd/0/ecp-copa/cabana/build/core/src/CMakeFiles/cabanacore.dir/depend.internal".
Scanning dependencies of target cabanacore
make[2]: Leaving directory `/root/hpc-gitlab-runner/ecp-copa/cabana/builds/users/junghans/47b77dfd/0/ecp-copa/cabana/build'
make -f core/src/CMakeFiles/cabanacore.dir/build.make core/src/CMakeFiles/cabanacore.dir/build
make[2]: Entering directory `/root/hpc-gitlab-runner/ecp-copa/cabana/builds/users/junghans/47b77dfd/0/ecp-copa/cabana/build'
[  2%] Building CXX object core/src/CMakeFiles/cabanacore.dir/impl/Cabana_Version.cpp.o
cd /root/hpc-gitlab-runner/ecp-copa/cabana/builds/users/junghans/47b77dfd/0/ecp-copa/cabana/build/core/src && /sw/ascent/gcc/6.4.0/bin/c++   -I/root/hpc-gitlab-runner/ecp-copa/cabana/builds/users/junghans/47b77dfd/0/ecp-copa/cabana/core/src -I/root/hpc-gitlab-runner/ecp-copa/cabana/builds/users/junghans/47b77dfd/0/ecp-copa/cabana/build/core/src -isystem /root/hpc-gitlab-runner/ecp-copa/cabana/builds/users/junghans/47b77dfd/0/ecp-copa/cabana/kokkos.install/include  -O3 -DNDEBUG   --std=c++11 -Xcudafe --diag_suppress=esa_on_defaulted_function_ignored -o CMakeFiles/cabanacore.dir/impl/Cabana_Version.cpp.o -c /root/hpc-gitlab-runner/ecp-copa/cabana/builds/users/junghans/47b77dfd/0/ecp-copa/cabana/core/src/impl/Cabana_Version.cpp
Scanning dependencies of target cabana_core_gtest
make[2]: Leaving directory `/root/hpc-gitlab-runner/ecp-copa/cabana/builds/users/junghans/47b77dfd/0/ecp-copa/cabana/build'
make -f core/unit_test/CMakeFiles/cabana_core_gtest.dir/build.make core/unit_test/CMakeFiles/cabana_core_gtest.dir/build
make[2]: Entering directory `/root/hpc-gitlab-runner/ecp-copa/cabana/builds/users/junghans/47b77dfd/0/ecp-copa/cabana/build'
c++: error: unrecognized command line option '-Xcudafe'
[  5%] Building CXX object core/unit_test/CMakeFiles/cabana_core_gtest.dir/__/__/gtest/gtest/gtest-all.cc.o
cd /root/hpc-gitlab-runner/ecp-copa/cabana/builds/users/junghans/47b77dfd/0/ecp-copa/cabana/build/core/unit_test && /sw/ascent/gcc/6.4.0/bin/c++   -I/root/hpc-gitlab-runner/ecp-copa/cabana/builds/users/junghans/47b77dfd/0/ecp-copa/cabana/gtest -I/root/hpc-gitlab-runner/ecp-copa/cabana/builds/users/junghans/47b77dfd/0/ecp-copa/cabana/build/core/unit_test -I/root/hpc-gitlab-runner/ecp-copa/cabana/builds/users/junghans/47b77dfd/0/ecp-copa/cabana/core/unit_test  -DGTEST_HAS_PTHREAD=0 -O3 -DNDEBUG   -std=c++11 -o CMakeFiles/cabana_core_gtest.dir/__/__/gtest/gtest/gtest-all.cc.o -c /root/hpc-gitlab-runner/ecp-copa/cabana/builds/users/junghans/47b77dfd/0/ecp-copa/cabana/gtest/gtest/gtest-all.cc
c++: error: unrecognized command line option '--diag_suppress=esa_on_defaulted_function_ignored'
make[2]: *** [core/src/CMakeFiles/cabanacore.dir/impl/Cabana_Version.cpp.o] Error 1
make[2]: Target `core/src/CMakeFiles/cabanacore.dir/build' not remade because of errors.
make[2]: Leaving directory `/root/hpc-gitlab-runner/ecp-copa/cabana/builds/users/junghans/47b77dfd/0/ecp-copa/cabana/build'
make[1]: *** [core/src/CMakeFiles/cabanacore.dir/all] Error 2
[  7%] Linking CXX static library libcabana_core_gtest.a
cd /root/hpc-gitlab-runner/ecp-copa/cabana/builds/users/junghans/47b77dfd/0/ecp-copa/cabana/build/core/unit_test && /autofs/nccsopen-svm1_sw/ascent/.swci/0-core/opt/spack/20180914/linux-rhel7-ppc64le/gcc-4.8.5/cmake-3.12.2-vltu2cerbdpbvzwpil2fvw4skvjmabfl/bin/cmake -P CMakeFiles/cabana_core_gtest.dir/cmake_clean_target.cmake
cd /root/hpc-gitlab-runner/ecp-copa/cabana/builds/users/junghans/47b77dfd/0/ecp-copa/cabana/build/core/unit_test && /autofs/nccsopen-svm1_sw/ascent/.swci/0-core/opt/spack/20180914/linux-rhel7-ppc64le/gcc-4.8.5/cmake-3.12.2-vltu2cerbdpbvzwpil2fvw4skvjmabfl/bin/cmake -E cmake_link_script CMakeFiles/cabana_core_gtest.dir/link.txt --verbose=1
/bin/ar qc libcabana_core_gtest.a  CMakeFiles/cabana_core_gtest.dir/__/__/gtest/gtest/gtest-all.cc.o
/bin/ranlib libcabana_core_gtest.a
make[2]: Leaving directory `/root/hpc-gitlab-runner/ecp-copa/cabana/builds/users/junghans/47b77dfd/0/ecp-copa/cabana/build'
[  7%] Built target cabana_core_gtest
make[1]: Target `all' not remade because of errors.
make[1]: Leaving directory `/root/hpc-gitlab-runner/ecp-copa/cabana/builds/users/junghans/47b77dfd/0/ecp-copa/cabana/build'
make: *** [all] Error 2
make: Target `default_target' not remade because of errors.
ERROR: Job failed: exit status 2
```

### PR DESCRIPTION
**Pull request by junghans**
Thanks to @jgalarowicz

Fix #67 

To Do:
- [x] re-enable the other backends

### CODE REVIEW

File: .ecp-gitlab-ci.yml [hunk 1]
```diff
@@ -49,62 +54,75 @@ stages:
...
-    BACKENDS: "OpenMP"
-  extends: .BuildKokkos
-
-Build OpenMP:
-  variables:
-    BACKENDS: "OpenMP"
-  extends: .Build
-  dependencies:
-   - BuildKokkos OpenMP
+#BuildKokkos Serial:
```
Comment (sslattery) [inline .ecp-gitlab-ci.yml]: Should we remove the commented code?

Comment (junghans) [reply .ecp-gitlab-ci.yml]: This is still work in progress, ultimately these will get reenabled!

**Comment by junghans** [comment]
Ready to merge @dalg24 please review!

**Comment by junghans** [comment]
@jgalarowicz, it works, but sometimes the tests fails with:
```
24/30 Test #24: Core_tutorial_04 .................***Exception: Child aborted  1.61 sec
terminate called after throwing an instance of 'std::runtime_error'
  what():  cudaDeviceSynchronize() error( cudaErrorDevicesUnavailable): all CUDA-capable devices are busy or unavailable /root/hpc-gitlab-runner/ecpcitest/ecp-copa/cabana/builds/users/junghans/2e8de492/1/ecpcitest/ecp-copa/cabana/kokkos/core/src/Cuda/Kokkos_Cuda_Impl.cpp:119
Traceback functionality not available
```
which means we should run the test as batch job instead.

What `tags:` do need to use to submit this to the queue?

**Comment by jgalarowicz** [comment]
@junghans I believe the tag  to submit a batch job to run on the compute nodes is "batch".

**Comment by junghans** [comment]
@jgalarowicz can you have a look why the last stage (`test`) is failing?

**Comment by jgalarowicz** [comment]
@junghans Yes, I will take a look!

**Comment by junghans** [comment]
Thanks, there is just no error message, which confuses me!

**Comment by jgalarowicz** [comment]
@junghans For some reason I can't log into ORNL.  I opened a ticket.  I will try this when I can login again.

**Comment by jgalarowicz** [comment]
@junghans My account at ORNL has been disabled.  I think because my INCITE PEAC allocation was not renewed.  I'm asking ORNL representatives for a sponsor.

**Comment by junghans** [comment]
I totally forgot about this PR!

@jgalarowicz it seems the permission issue is back:
https://code.ornl.gov/ecpcitest/ecp-copa/cabana/pipelines/42620
can you have a look?

**Comment by jgalarowicz** [comment]
@junghans  It seems like this might be the problem where each of the tests need to be a separate stage?    I see that the code from 1ab6b95 that was the initial try on this.  But, I don't see that code in the repository now.  I remember you saying it wouldn't scale because of all the different variations that are required.

**Comment by jgalarowicz** [comment]
@junghans - I see the code now in the ci-cuda branch.   So, maybe a different issue.  Consulting with NMC - Paul and others.

**Comment by sslattery** [comment]
What is the status of this PR? Are we still seeing issues?

**Comment by junghans** [comment]
@jgalarowicz is doing final tweaks!

**Comment by junghans** [comment]
@jgalarowicz I added a workaround for the serialization bug.

**Comment by junghans** [comment]
This works now: https://code.ornl.gov/ecpcitest/ecp-copa/cabana/pipelines/45674

@sslattery @dalg24 please review and merge.

File: .ecp-gitlab-ci.yml [hunk 2]
```diff
@@ -33,19 +35,66 @@ stages:
...
-    - module load gcc
     - for i in ${BACKENDS}; do CMAKE_OPTS+=( -DCabana_ENABLE_${i}=ON ); done
     - j="$(grep -c processor /proc/cpuinfo 2>/dev/null)" || j=0; ((j++))
     - mkdir build && cd build &&
-      cmake -DCMAKE_PREFIX_PATH=${CI_PROJECT_DIR}/kokkos.install -DCabana_ENABLE_TESTING=ON -DCabana_ENABLE_Serial=OFF -DCabana_ENABLE_EXAMPLES=OFF ${CMAKE_OPTS[@]} .. &&
-      make -k -j${j} VERBOSE=1 &&
-      make test CTEST_OUTPUT_ON_FAILURE=1
+      cmake -DCMAKE_PREFIX_PATH=${CI_PROJECT_DIR}/kokkos.install -DCabana_ENABLE_TESTING=ON -DCabana_ENABLE_Serial=OFF -DCabana_ENABLE_EXAMPLES=ON ${CMAKE_OPTS[@]} ${CMAKE_EXTRA_ARGS} .. &&
+      make -k -j${j} VERBOSE=1
+    - wget https://raw.githubusercontent.com/slashbeast/pax-utils/master/lddtree.sh
```
Comment (sslattery) [inline .ecp-gitlab-ci.yml]: What is this here for?

File: .ecp-gitlab-ci.yml [hunk 3]
```diff
@@ -33,19 +35,66 @@ stages:
...
+    - module load gcc/6.4.0
+    - module load cuda
+
+.Test_Cuda:
+  extends: .Test
+  before_script:
+    - module load gcc/6.4.0
+    - module load cuda
+
+BuildKokkos Cuda:
```
Comment (sslattery) [inline .ecp-gitlab-ci.yml]: Is this extra hierarchy needed here? Why not just use `BuildKokkos_Cuda` directly?

Comment (junghans) [reply .ecp-gitlab-ci.yml]: Good point, it is not needed, but I had mulitple Cuda builds with different options in mind.

File: .ecp-gitlab-ci.yml (hunk 2 above)
Comment (junghans) [reply .ecp-gitlab-ci.yml]: Removed.

**Comment by junghans** [comment]
@sslattery squashed and rebased.

File: .ecp-gitlab-ci.yml (hunk 3 above)
Comment (junghans) [reply .ecp-gitlab-ci.yml]: And I wanted to keep it similar to the other backends.

File: .ecp-gitlab-ci.yml (hunk 2 above)
Comment (sslattery) [reply .ecp-gitlab-ci.yml]: it is still there...

File: .ecp-gitlab-ci.yml (hunk 3 above)
Comment (sslattery) [reply .ecp-gitlab-ci.yml]: So do you want to keep this then or should we remove the extra layer for now and add it back when we have a need for additional build arrangements?

Comment (junghans) [reply .ecp-gitlab-ci.yml]: I would like to keep it.

File: .ecp-gitlab-ci.yml (hunk 2 above)
Comment (sslattery) [reply .ecp-gitlab-ci.yml]: I guess I just wanted to know what the `wget` call did in this case

Comment (junghans) [reply .ecp-gitlab-ci.yml]: That should be removed! I must have missed that in the squash.

Comment (junghans) [reply .ecp-gitlab-ci.yml]: Now removed.

**Pull request merged**